src/items/appItemHooks.ts

```diff
@@ -0,0 +1,143 @@
+import { useEffect, useState } from 'react';
+
+import { Context, DEFAULT_LANG, UUID } from '@graasp/sdk';
+import { ItemRecord, MemberRecord } from '@graasp/sdk/frontend';
+
+export type Token = string;
+
+const buildPostMessageKeys = (
+  itemId: UUID,
+): {
+  [key: string]: string;
+} => ({
+  GET_CONTEXT_SUCCESS: `GET_CONTEXT_SUCCESS_${itemId}`,
+  GET_CONTEXT_FAILURE: `GET_CONTEXT_FAILURE_${itemId}`,
+  GET_CONTEXT: `GET_CONTEXT_${itemId}`,
+  GET_AUTH_TOKEN: `GET_AUTH_TOKEN_${itemId}`,
+  GET_AUTH_TOKEN_SUCCESS: `GET_AUTH_TOKEN_SUCCESS_${itemId}`,
+  GET_AUTH_TOKEN_FAILURE: `GET_AUTH_TOKEN_FAILURE_${itemId}`,
+  POST_AUTO_RESIZE: `POST_AUTO_RESIZE_${itemId}`,
+});
+
+const useAppCommunication = ({
+  item,
+  member,
+  appUrl,
+  apiHost,
+  permission,
+  context,
+  iFrameRef,
+  requestApiAccessToken,
+}: {
+  item: ItemRecord;
+  member?: MemberRecord;
+  appUrl: string;
+  apiHost: string;
+  permission: string;
+  context?: `${Context}` | Context;
+  iFrameRef: React.RefObject<HTMLIFrameElement>;
+  requestApiAccessToken: (payload: {
+    id: UUID;
+    app: string;
+    origin: string;
+  }) => Promise<{
+    token: Token;
+  }>;
+}): void => {
+  const [channel, setChannel] = useState<MessageChannel>();
+
+  useEffect(() => {
+    // receive message from app through MessageChannel
+    const onMessage = async (e: MessageEvent): Promise<void> => {
+      const { data, origin: requestOrigin } = e;
+
+      const POST_MESSAGE_KEYS = buildPostMessageKeys(item.id);
+
+      // responds only to corresponding app
+      if (!appUrl?.includes(requestOrigin)) {
+        return;
+      }
```

Comment:
We could check for `source` as well, to ensure that it comes from the iframe (https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage#the_dispatched_event)

Reply:
In my tests the `source` is `null`

Reply:
Ah right because of the sandboxing